### Installing and Importing Required Libraries

In [2]:
#Importing Required Libraries
import cvxpy as cp
import pandas as pd
import numpy as np

1.

For plant A:

**Decision variables:**

\$X_1\$ = Number of standard units

\$X_2\$ = Number of deluxe units

The objective function is:
\begin{equation}
\text{maximize}_{X1,X2}\quad 10 \times X1 + 15 \times X2
\end{equation}
subject to constraints
\begin{align}
4X1 + 2X2 &\leq 80&\
2X1 + 5X2 &\leq 60&\
4X1 + 4X2 &\leq 75&\
X1, X2 &\geq 0&
X1, X2 , \in \mathbb{Z}&
\end{align}

For plant B:

**Decision variables:**

\$Y_1\$ = Number of standard units

\$Y_2\$ = Number of deluxe units

The objective function is:
\begin{equation}
\text{maximize}_{Y1,Y2}\quad 10 \times Y1 + 15 \times Y2
\end{equation}
subject to constraints
\begin{align}
5Y1 + 3Y2 &\leq 60&\
5Y1 + 6Y2 &\leq 75&\
4Y1 + 4Y2 &\leq 45&\
Y1, Y2 &\geq 0&
Y1, Y2 , \in \mathbb{Z}&
\end{align}

### 2. Individual Optimization Models

In [ ]:
#Creating a function to calculate the optimal solutions and profit levels for each plant
def individual_plant():
    
    #Defining the variables as integers - subject to the constraint of being integers
    x1 = cp.Variable(integer=True)
    x2 = cp.Variable(integer=True)
    y1 = cp.Variable(integer=True)
    y2 = cp.Variable(integer=True)

    #Defining constraints for Plant A
    constraints_a = [
        4 * x1 + 2 * x2 <= 80,
        2 * x1 + 5 * x2 <= 60,
        4 * (x1 + x2) <= 75,
        x1 >= 0,
        x2 >= 0
    ]

    #Defining constraints for Plant B
    constraints_b = [
        5 * y1 + 3 * y2 <= 60,
        5 * y1 + 6 * y2 <= 75,
        4 * (y1 + y2) <= 45,
        y1 >= 0,
        y2 >= 0
    ]

    #Defining the objective function
    obj_a = cp.Maximize(10 * x1 + 15 * x2)
    obj_b = cp.Maximize(10 * y1 + 15 * y2)

    #Solving the problem
    prob_a = cp.Problem(obj_a, constraints_a)
    prob_b = cp.Problem(obj_b, constraints_b)
    prob_a.solve()
    prob_b.solve()

    #Printing the results
    print("Plant A optimal solution:")
    print("Standard units: ", x1.value)
    print("Deluxe units: ", x2.value)
    print("Maximized Profit: ", prob_a.value)

    print("\nPlant B optimal solution:")
    print("Standard units: ", y1.value)
    print("Deluxe units: ", y2.value)
    print("Maximized Profit: ", prob_b.value)

#Solving individual plant problems
individual_plant()

Plant A optimal solution:
Standard units:  10.0
Deluxe units:  8.0
Maximized Profit:  220.0

Plant B optimal solution:
Standard units:  0.0
Deluxe units:  11.0
Maximized Profit:  165.0


We observe that when calculating optimal values for each plant individually, 

|  | Plant A | Plant B         
| :---:|:---: | :---:
| Standard Units|10|0
| Deluxe Units|8|11
| Maximized Profit|220|165

Plant A focuses on Standard and Deluxe observes a profit of $220.

Plant B focses only on Deluxe Units and observes a profit of $165.

### 3. Joint Optimization Model

In [ ]:
#Creating a function to calculate the optimal solution and profit level for both plants combined
def joint_plant():

    #Defining the variables as integers - subject to the constraint of being integers
    x1 = cp.Variable(integer=True)
    x2 = cp.Variable(integer=True)
    y1 = cp.Variable(integer=True)
    y2 = cp.Variable(integer=True)

    #Defining the combined list of constraints
    constraints = [
        4 * x1 + 2 * x2 <= 80,
        2 * x1 + 5 * x2 <= 60,
        5 * y1 + 3 * y2 <= 60,
        5 * y1 + 6 * y2 <= 75,
        4 * (x1 + x2 + y1 + y2) <= 120,
        x1 >= 0,
        x2 >= 0,
        y1 >= 0,
        y2 >= 0
    ]

    #Defining the objective function
    obj = cp.Maximize(10 * x1 + 15 * x2 + 10 * y1 + 15 * y2)

    #Solving the problem
    prob = cp.Problem(obj, constraints)
    prob.solve()

    #Printing the results
    print("Joint optimal solution:")
    print("Plant A - Standard units: ", x1.value)
    print("Plant A - Deluxe units: ", x2.value)
    print("Plant B - Standard units: ", y1.value)
    print("Plant B - Deluxe units: ", y2.value)
    print("Maximized Profit: ", prob.value)

#Solving the joint plant problem
joint_plant()

Joint optimal solution:
Plant A - Standard units:  10.0
Plant A - Deluxe units:  8.0
Plant B - Standard units:  0.0
Plant B - Deluxe units:  12.0
Maximized Profit:  400.0


We observe that when calculating optimal values for each plant individually, 

|  | Plant A | Plant B         
| :---:|:---: | :---:
| Standard Units|10|0
| Deluxe Units|8|12

Plant A focuses on Standard and Deluxe Units.

Plant B focses only on Deluxe Units.

A maximum profit of $400 can be attained.

### 4.
The results show that optimizing each plant individually does not necessarily lead to the overall optimal solution when considering the system as a whole. In the individual optimization case, plant B produces 11 deluxe units leading to a profit of \$165. However, in the joint optimization case, while plant A's production remains the same, plant B shifts its production to produce an additional deluxe unit, resulting in an overall profit of \$400. This is an increase of \$15 compared to the sum of the profits from the individual optimizations (\$220 + \$165 = \$385). This highlights the importance of considering the entire system when making optimization decisions, as local optimizations can lead to suboptimal results when considering the system as a whole.

### 5.
The optimization analysis indicates that while Plant A's production levels remain optimal, Plant B could enhance profitability by increasing its production of deluxe units. Therefore, my foremost recommendation would be to review and revise the production schedules and resource allocation between plants A and B, focusing on maximizing the production of deluxe units at Plant B. 

Notably, the optimization model also suggests that the older machinery at Plant B, resulting in longer processing times, could be limiting its production efficiency. Upgrading or replacing this machinery might yield significant improvements in the plant's capacity and efficiency, potentially increasing profitability. This, however, should be evaluated against the cost and feasibility of such an investment.

Moreover, it might be worthwhile to explore options for increasing the raw material supply or finding alternatives. The raw material constraint significantly impacts the production potential of both plants. 

Finally, this analysis underscores the crucial importance of a system-wide view for decision-making. Optimizing operations at the individual plant level does not necessarily yield the best overall results. As such, future operational strategies should be based on a comprehensive, holistic assessment of the entire production system, incorporating all relevant factors, including but not limited to production capacities, raw material availability, demand projections, and operational costs.